## В этом проекте посмотрим на общую статистику по отделу саппорт за 2020 год. Узнаем медиану времени первого ответа по часам в будни и выходные, рассмотрим гипотезу о том, что время первого ответа клиентам влияет на пользовательский рейтинг.
## Для февраля 2021 узнаем какое количество обращений приходится на утреннюю, вечернюю смены и их перекрестие, а также количество обращений в нерабочие часы и медиану времени ожидания ответа. Посмотрим на соотношение обращений от юзеров и лидов.


In [1]:
import pandas as pd
import plotly.express as px
from scipy import stats as st
import numpy as np

In [2]:
intercom2020 = pd.read_csv('inbox-data-export.csv')

In [3]:
intercom2020.head()

Conversation ID  ...   Source
0   19131500062955  ...  Article
1   19131500063032  ...      NaN
2   19131500062837  ...      NaN
3   19131500062826  ...      NaN
4   19131500062661  ...      NaN

[5 rows x 27 columns]

In [4]:
intercom2020['Minutes to first reply'] = intercom2020['Time to first reply (seconds)'] / 60

In [5]:
intercom2020['Time Zone'].unique()

array(['London'], dtype=object)

# Посмотрим на количество обращений по каналам связи.

In [6]:
channels = intercom2020.pivot_table(index='Channel', values='Conversation ID', aggfunc='count')
channels = channels.reset_index()
channels

Channel  Conversation ID
0  Android App             3424
1      Desktop            27103
2        Email             3259
3     Facebook              113
4      iOS App             4064

In [7]:
px.pie(channels, names='Channel',values='Conversation ID')

In [8]:
intercom2020['Created at'] = pd.to_datetime(intercom2020['Created at'])

In [9]:
intercom2020['Last updated at'] = pd.to_datetime(intercom2020['Created at'])

In [10]:
by_source = intercom2020.pivot_table(index='Source', values='Conversation ID', aggfunc='count')
by_source

Conversation ID
Source                            
Article                        247
Manual message                 425
User auto message             8528

In [12]:
intercom2020.describe()

Conversation ID      Reopened  ...  Teammate replies  Minutes to first reply
count     3.796300e+04  37963.000000  ...      37963.000000            28802.000000
mean      1.029124e+13      0.035772  ...          2.953507               81.066132
std       9.526151e+12      0.205774  ...          4.656707             3696.201216
min       2.522102e+10      0.000000  ...          0.000000                0.033333
25%       2.623905e+10      0.000000  ...          1.000000                3.816667
50%       1.913150e+13      0.000000  ...          1.000000                9.683333
75%       1.913150e+13      0.000000  ...          3.000000               39.245833
max       1.913150e+13      5.000000  ...        114.000000           614236.733333

[8 rows x 8 columns]

In [13]:
px.box(intercom2020, y='Minutes to first reply')

In [48]:
#Удалим выбросы
q995 = intercom2020['Minutes to first reply'].quantile(0.995)

In [15]:
intercom2020 = intercom2020[intercom2020['Minutes to first reply'] <= q995]
intercom2020 = intercom2020.reset_index()

# Изучим медиану времени первого ответа клиентам на буднях и на выходных

In [16]:
px.histogram(intercom2020, x='Minutes to first reply', range_x=(0,100))

In [19]:
intercom2020['Created_at_hour'] = intercom2020['Created at'].dt.hour

In [20]:
intercom2020['Created_at_day'] = intercom2020['Created at'].dt.dayofweek

In [21]:
intercom2020['type_of_day'] =  ['Weekend' if i > 4 else 'Business Day' for i in intercom2020['Created_at_day']]
#intercom2020

In [22]:
pivot = intercom2020.pivot_table(index=['type_of_day', 'Created_at_hour'], aggfunc={'Minutes to first reply':['mean','median']})
pivot = pivot.reset_index()

In [23]:
pivot.columns = pivot.columns.droplevel()

In [24]:
pivot.columns = ['type_of_day', 'hour', 'mean', 'median']
#pivot

In [25]:
temp = pivot[pivot['type_of_day'] == 'Business Day']
px.bar(temp, x = 'hour', y ='median', title = 'Business Day Median First Response Time') # 8 AM-7 PM UTC working hours

In [26]:
temp2 = pivot[pivot['type_of_day'] == 'Weekend']
px.bar(temp, x = 'hour', y ='mean', title = 'Weekend Median First Response Time') # 9 AM-5 PM working hours

# Проверим гипотезу о том, что время ответа больше 10 минут влияет на пользовательский рейтинг

In [27]:
group_fast = intercom2020[intercom2020['Minutes to first reply'] < 10].dropna(subset=['Conversation rating']).reset_index()
group_slow = intercom2020[intercom2020['Minutes to first reply'] > 10].dropna(subset=['Conversation rating']).reset_index()

За нулевую гипотезу возьмем утверждение, что время ответа на пользовательский рейтинг не влияет. За альтернативную, что влияет.

In [28]:
results = st.mannwhitneyu(group_fast['Conversation rating'], group_slow['Conversation rating'])

if results.pvalue < 0.05:
  print('Нулевая гипотеза может быть отвергнута')
else:
  print('Нулевая гипотеза не может быть отвергнута')

round(group_fast['Conversation rating'].mean() -  group_slow['Conversation rating'].mean(), 3)

Нулевая гипотеза может быть отвергнута


0.125

По результатам теста видим, что оценки у чатов со временем ответа больше и меньше 10 минут различны. 
Также, видим, что средняя оценка диалогов со временем первого ответа меньше 10 минут выше на 0.125.
Отсюда можем сделать вывод, что время первого ответа клиенту меньше 10 минут влияет на пользовательский рейтинг. 

В следующем тесте за нулевую гипотезу возьмем то, что время первого ответа клиенту не различается у групп с высоким и низким пользовательским рейтингом. За альтернативную возьмем, что время ответа у двух групп различается.

In [29]:
group_high_rate = intercom2020[intercom2020['Conversation rating'] > 3].dropna(subset=['Minutes to first reply']).reset_index()
group_low_score = intercom2020[intercom2020['Conversation rating'] < 3].dropna(subset=['Minutes to first reply']).reset_index()

In [30]:
results = st.mannwhitneyu(group_high_rate['Minutes to first reply'], group_low_score['Minutes to first reply'])

if results.pvalue < 0.05:
  print('Нулевая гипотеза может быть отвергнута')
else:
  print('Нулевая гипотеза не может быть отвергнута')

round(group_low_score['Minutes to first reply'].mean() -  group_high_rate['Minutes to first reply'].mean(), 1)

Нулевая гипотеза может быть отвергнута


9.2

Видим, что среднее время ответа в двух группах различно. Также, среднее время ответа в группе диалогов с низким рейтингом выше на 9.2 минуты.

# Узнаем как менялось количество обращений и среднее время ответа в течение дня в феврале

In [31]:
#таблица с количеством человек на сменах по дням
february_shifts = pd.read_csv('/content/timetable_feb.csv')
february_shifts.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28 entries, 0 to 27
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   День    28 non-null     object
 1   утро    28 non-null     int64 
 2   вечер   28 non-null     int64 
dtypes: int64(2), object(1)
memory usage: 800.0+ bytes


In [32]:
february_shifts['День'] = pd.to_datetime(february_shifts['День'], format = '%m/%d/%Y')

In [33]:
#выгрузка из интеркома за февраль
intercom_feb = pd.read_csv('/content/Intercom_feb.csv')
intercom_feb

Conversation ID  ...             Source
0      19131500138140  ...  User auto message
1      19131500136713  ...                NaN
2      19131500137984  ...                NaN
3      19131500137306  ...                NaN
4      19131500131163  ...                NaN
...               ...  ...                ...
7071   19131500138478  ...                NaN
7072   19131500138479  ...                NaN
7073   19131500138480  ...                NaN
7074   19131500138481  ...                NaN
7075   19131500138894  ...                NaN

[7076 rows x 23 columns]

In [34]:
intercom_feb['Created at'] = pd.to_datetime(intercom_feb['Created at'])

In [35]:
intercom_feb['month'] = intercom_feb['Created at'].dt.month

In [36]:
intercom_feb['Created_at_hour'] = intercom_feb['Created at'].dt.hour

In [37]:
intercom_feb['Minutes to first reply'] = intercom_feb['Time to first reply (seconds)'] / 60

In [38]:
intercom_feb['Date_only'] = intercom_feb['Created at'].dt.date
intercom_feb

Conversation ID  ...   Date_only
0      19131500138140  ...  2021-02-24
1      19131500136713  ...  2021-02-23
2      19131500137984  ...  2021-02-24
3      19131500137306  ...  2021-02-23
4      19131500131163  ...  2021-02-17
...               ...  ...         ...
7071   19131500138478  ...  2021-02-25
7072   19131500138479  ...  2021-02-25
7073   19131500138480  ...  2021-02-25
7074   19131500138481  ...  2021-02-25
7075   19131500138894  ...  2021-02-25

[7076 rows x 27 columns]

In [39]:
intercom_feb = intercom_feb.set_index('Date_only')
february_shifts = february_shifts.set_index('День')


In [40]:
#объединим таблицу из интеркома и таблицу с кол-вом человек на смене
feb_merged = pd.merge(intercom_feb, february_shifts, how='right', right_index=True, left_index=True)

In [41]:
#зададим временные промежутки (утро, пересечение смен, вечер и нерабочие часы)
feb_merged['session'] = ['morning' if i in [8,9,10] else 'both' if i in [11,12,13,14,15] \
                         else 'evening' if i in [16,17,18] else 'non working' for i in feb_merged['Created_at_hour']]
feb_merged

Conversation ID  ...      session
2021-02-01   19131500118186  ...      evening
2021-02-01   19131500118213  ...      evening
2021-02-01   19131500118392  ...  non working
2021-02-01   19131500118439  ...  non working
2021-02-01   19131500118056  ...         both
...                     ...  ...          ...
2021-02-28   19131500140921  ...      evening
2021-02-28   19131500140946  ...      evening
2021-02-28   19131500141040  ...  non working
2021-02-28   19131500140954  ...      evening
2021-02-28   19131500140992  ...  non working

[7038 rows x 29 columns]

## Найдем количество входящих обращений в каждом периоде

In [42]:
shift_pivot = feb_merged.pivot_table(index='session', values='Conversation ID', aggfunc='count').reset_index()
shift_pivot

session  Conversation ID
0         both             2624
1      evening             1378
2      morning             1236
3  non working             1800

In [43]:
px.bar(shift_pivot, x = 'session', y ='Conversation ID', title = 'Conversations during the day')

## Найдем среднее и медиану времени первого ответа в каждом периоде

In [44]:
mfrt_by_shift = feb_merged.pivot_table(index='session', values='Minutes to first reply', aggfunc=['mean', 'median']).reset_index()
mfrt_by_shift.columns = mfrt_by_shift.columns.droplevel()
mfrt_by_shift.columns = ['Shifts', 'Mean FRT', 'Median FRT']
mfrt_by_shift

Shifts    Mean FRT  Median FRT
0         both  154.719509   12.300000
1      evening  165.733402   17.608333
2      morning  193.860941   53.533333
3  non working  305.273957  106.200000

In [45]:
px.bar(mfrt_by_shift, x = 'Shifts', y ='Median FRT', title = 'Median FRT during the day')

# Посмотрим на количество обращений от юзеров и от лидов и медиану времени первого ответа

In [46]:
user_lead_pivot = intercom_feb.pivot_table(index='Type', aggfunc={'Conversation ID': 'count', 'Minutes to first reply': 'median',\
                                                                  'Teammate replies':'mean'}).reset_index()
user_lead_pivot

Type  Conversation ID  Minutes to first reply  Teammate replies
0  Lead             1072               30.025000          1.408582
1  User             6004               35.383333          2.426716

In [49]:
pivot_replied = feb_merged.pivot_table(index=['Teammate replies', 'Type'], values='Conversation ID', aggfunc='count').reset_index()
pivot_replied

Teammate replies  Type  Conversation ID
0                  0  Lead              360
1                  0  User             1655
2                  1  Lead              431
3                  1  User             1743
4                  2  Lead              133
5                  2  User              827
6                  3  Lead               51
7                  3  User              465
8                  4  Lead               36
9                  4  User              324
10                 5  Lead               12
11                 5  User              252
12                 6  Lead               18
13                 6  User              155
14                 7  Lead                5
15                 7  User              131
16                 8  Lead                4
17                 8  User               88
18                 9  Lead                2
19                 9  User               75
20                10  Lead                5
21                10  User               43
22                11  Lead                3
23                11  User               54
24                12  Lead                3
25                12  User               26
26                13  User               23
27                14  User               16
28                15  User               16
29                16  Lead                1
30                16  User               11
31                17  Lead                1
32                17  User                8
33                18  User                7
34                19  User                4
35                20  Lead                1
36                20  User                7
37                21  Lead                2
38                21  User                6
39                22  Lead                1
40                22  User                1
41                23  Lead                1
42                23  User                1
43                24  User                6
44                25  User                1
45                26  User                3
46                27  User                1
47                28  User                3
48                30  User                3
49                31  User                3
50                34  User                1
51                35  User                2
52                42  User                1
53                43  User                1
54                44  User                2
55                57  User                1
56                65  User                1
57                74  User                1

In [52]:
px.bar(pivot_replied, x = 'Teammate replies', y ='Conversation ID', title = 'Median FRT during the day', color='Type')